server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java

```diff
@@ -80,7 +80,12 @@ public void run() throws IOException {
 
                     @Override
                     protected void innerOnResponse(QuerySearchResult response) {
-                        counter.onResult(response);
+                        try {
+                            counter.onResult(response);
+                        } catch (Exception e) {
+                            assert false : e;
```

Comment:
I don't think we should have an assert here. An exception could happen in a test but the assert won't be of any help since this fix will propagate the appropriate error (phase failure) to the caller.